src/main/java/net/minecraftforge/items/wrapper/EntityEquipmentInvWrapper.java

```diff
@@ -0,0 +1,170 @@
+
+        int limit = getStackLimit(slot, stack);
+
+        if (!existing.isEmpty())
+        {
+            if (!ItemHandlerHelper.canItemStacksStack(stack, existing))
+                return stack;
+
+            limit -= existing.getCount();
+        }
+
+        if (limit <= 0)
+            return stack;
+
+        boolean reachedLimit = stack.getCount() > limit;
+
+        if (!simulate)
+        {
+            if (existing.isEmpty())
+            {
+                entity.setItemStackToSlot(equipmentSlot, reachedLimit ? ItemHandlerHelper.copyStackWithSize(stack, limit) : stack);
+            }
+            else
+            {
+                existing.grow(reachedLimit ? limit : stack.getCount());
+            }
+        }
+
+        return reachedLimit ? ItemHandlerHelper.copyStackWithSize(stack, stack.getCount() - limit) : ItemStack.EMPTY;
+    }
+
+    @Nonnull
+    @Override
+    public ItemStack extractItem(final int slot, final int amount, final boolean simulate)
+    {
+        if (amount == 0)
+            return ItemStack.EMPTY;
+
+        final EntityEquipmentSlot equipmentSlot = validateSlotIndex(slot);
+
+        final ItemStack existing = entity.getItemStackFromSlot(equipmentSlot);
+
+        if (existing.isEmpty())
+            return ItemStack.EMPTY;
+
+        final int toExtract = Math.min(amount, existing.getMaxStackSize());
+
+        if (existing.getCount() <= toExtract)
+        {
+            if (!simulate)
+            {
+                entity.setItemStackToSlot(equipmentSlot, ItemStack.EMPTY);
+            }
+
+            return existing;
+        }
+        else
+        {
+            if (!simulate)
+            {
+                entity.setItemStackToSlot(equipmentSlot, ItemHandlerHelper.copyStackWithSize(existing, existing.getCount() - toExtract));
+            }
+
+            return ItemHandlerHelper.copyStackWithSize(existing, toExtract);
+        }
+    }
+
+    @Override
+    public int getSlotLimit(final int slot)
+    {
+        final EntityEquipmentSlot equipmentSlot = validateSlotIndex(slot);
+        return equipmentSlot.getSlotType() == EntityEquipmentSlot.Type.ARMOR ? 1 : 64;
+    }
+
+    protected int getStackLimit(final int slot, @Nonnull final ItemStack stack)
+    {
+        return Math.min(getSlotLimit(slot), stack.getMaxStackSize());
+    }
+
+    @Override
+    public void setStackInSlot(final int slot, @Nonnull final ItemStack stack)
+    {
+        final EntityEquipmentSlot equipmentSlot = validateSlotIndex(slot);
+        if (ItemStack.areItemStacksEqual(entity.getItemStackFromSlot(equipmentSlot), stack))
+            return;
+        entity.setItemStackToSlot(equipmentSlot, stack);
+    }
+
+    protected EntityEquipmentSlot validateSlotIndex(final int slot)
+    {
+        if (slot < 0 || slot >= slots.size())
+            throw new RuntimeException("Slot " + slot + " not in valid range - [0," + slots.size() + ")");
```

Comment:
I just copied this method from `ItemStackHandler`, but throwing `IllegalArgumentException` makes sense.